main.go

```diff
@@ -0,0 +1,186 @@
+package main
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"net/http"
+	"os"
+	"strings"
+
+	"gopkg.in/yaml.v2"
+)
+
+type ConfigSettings struct {
+	Port  string   `yaml:"port"`
+	Repos []string `yaml:"repos"`
+}
+
+// Global Configuration for the service
+var Config ConfigSettings
+
+func main() {
+	var err error
+	configFile := "config.yaml"
+	if len(os.Args) == 1 {
+		fmt.Println("no configuration file specified, defaulting to config.yaml")
+	}
+	if len(os.Args) > 1 {
+		configFile = os.Args[1]
+	}
+
+	// Get the config file on startup to prevent repeated IO/ parsing
+	yamlData, err := ioutil.ReadFile(configFile)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	Config = ConfigSettings{}
+
+	err = yaml.Unmarshal(yamlData, &Config)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	if len(Config.Port) == 0 {
+		fmt.Println("no port specified in the yaml config file, defaulting to 8000")
+		Config.Port = "8000"
+	}
+
+	if len(Config.Repos) == 0 {
+		log.Fatal(errors.New("no git repos were configured in the yaml config file, at least one repo must be listed for the service to query"))
+	}
+
+	http.HandleFunc("/search", func(w http.ResponseWriter, r *http.Request) {
+		var err error
+
+		if r.Method != "GET" {
+			errorResp(w, fmt.Sprintf("incorect request method %s only the GET method is allowed", r.Method))
+			return
+		}
+
+		req := &SearchRequest{}
+		req.SearchTerm = r.URL.Query().Get("q")
+		if len(req.SearchTerm) == 0 {
+			errorResp(w, "no search term was found, query string must have a 'q' parameter which must be at least 1 character long")
+			return
+		}
+		req.User = r.URL.Query().Get("user")
+
+		searchResp, err := search(req)
+		if err != nil {
+			errorResp(w, fmt.Sprintf("search query could not be compleated, %s", err))
+			return
+		}
+
+		respJSON, err := json.Marshal(searchResp)
+		if err != nil {
+			errorResp(w, fmt.Sprintf("could not marshal query response %s", err))
+			return
+		}
+
+		w.Header().Add("Content-Type", "application/json")
+		w.Write(respJSON)
+	})
+
+	fmt.Println("starting service on port :" + Config.Port)
+	log.Fatal(http.ListenAndServe(":"+Config.Port, nil))
+}
+
+type SearchRequest struct {
+	SearchTerm string `json:"SearchTerm"`
+	User       string `json:"User"`
+}
+
+type Result struct {
+	FileURL string `json:"fileURL"`
+	Repo    string `json:"repo"`
+}
+
+type SearchResponse struct {
+	Results []Result
```

Comment:
Answer: After reviewing the code I believe it would be better for this to be a slice of struct pointers. It's quite possible that the Result struct could grow making a struct pointer increasingly more efficient. More importantly, the GitHub search API says it includes up to 1,000 results for each search. Storing pointers to these structs will probably be more efficient and could potentially prevent thousands of struct copies as this package grows and changes.